### PR TITLE
DAOS-3776 test: Adding test to verify aggregation across multiple pools and containers

### PR DESCRIPTION
**Pull request by saurabhtandan**
DAOS-3776 test: Adding test to verify aggregation across multiple pools and containers

As part of this test, existing file io/parallel.py
has been modified to add a test to verify aggregation
across multiple pools and containers.

Test-tag-hw-medium:pr,hw,medium,ib2 multipoolparallelio

Signed-off-by: Saurabh Tandan <saurabh.tandan@intel.com>

### CODE REVIEW

File: src/tests/ftest/io/parallel_io.py [hunk 1]
```diff
@@ -192,3 +239,157 @@ def test_parallelio(self):
...
+        Use cases:
+            Create 10 pools
+            Create 10 containers under each pool.
+            Record statvfs free space for each pool.
+            Perform parallel io to each pool without deleting the file
+            after write.
+            Record free space using statvfs after write.
+            Delete half of the containers from each pool.
+            Calculate the expected amount of data to be deleted when
+            containers are destroys.
```
Comment (shimizukko) [inline src/tests/ftest/io/parallel_io.py:258]:
```suggestion
            containers are destroyed.
```

File: src/tests/ftest/io/parallel_io.py [hunk 2]
```diff
@@ -192,3 +239,157 @@ def test_parallelio(self):
...
+                                 ret_code.items() if code != 0]))
+                        raise CommandFailure(
+                            "Error running '{}' on the following "
+                            "hosts: {}".format(cmd, error_hosts))
+                # report error if any command fails
+                except CommandFailure as error:
+                    self.log.error("ParallelIo Test Failed: %s",
+                                   str(error))
+                    self.fail("Test was expected to pass but "
+                              "it failed.\n")
```
Comment (shimizukko) [inline src/tests/ftest/io/parallel_io.py:329]: I see similar code in your PR3124. Maybe you can make it to a common method.

File: src/tests/ftest/io/parallel_io.py [hunk 3]
```diff
@@ -102,6 +123,42 @@ def start_dfuse(self):
...
+
+          Args:
+            path (str): path to get free block size of.
+          Returns:
+            integer value of stat free blocks
+        """
+        cmd = "ssh" " {}@{}" " stat -c%a -f {}".format(
+            getuser(), self.hostlist_clients[0], path)
+        try:
+            result = subprocess.check_output(cmd, shell=True)
```
Comment (dinghwah) [inline src/tests/ftest/io/parallel_io.py:137]: subprocess call with shell=True will trigger a high-severity security issue (will be caught by bandit).
Please provide a "list" for cmd, and remove the "shell=True".
Please see lanuch.py as reference.

File: src/tests/ftest/io/parallel_io.py [hunk 4]
```diff
@@ -192,3 +241,139 @@ def test_parallelio(self):
...
+
+        # Calculate the expected space to be returned after containers
+        # are destroyed.
+        reduced_space = (self.cont_count *
+                         int(self.ior_cmd.block_size.value))/2
+
+        # Verify if expected space is returned for each pool after containers
+        # were destroyed. If not, wait for 60 secs and check again. Wait 4
+        # times, otherwise exit the test with a failure.
+        for count in range(self.pool_count):
```
Comment (Puspalogan) [inline src/tests/ftest/io/parallel_io.py:362]: In the worst case, if something goes wrong, this for loop is gonna wait for 10 * 60 * 4 seconds. Why not run this as threads for each pool as well ?

File: src/tests/ftest/io/parallel_io.py [hunk 5]
```diff
@@ -192,3 +241,139 @@ def test_parallelio(self):
...
+
+        # Record free space after io
+        statvfs_before_cont_destroy = self.statvfs_pool(
+            self.dfuse.mount_dir.value)
+
+        # Destroy half of the containers from each pool
+        pfinal = 0
+        for count in range(self.cont_count):
+            pinitial = pfinal
+            pfinal = pinitial + (self.cont_count / 2)
```
Comment (Puspalogan) [inline src/tests/ftest/io/parallel_io.py:339]: Word of caution. If we upgrade to python 3.x, we need to be careful about how we use divide by "/". It is going to return floating point if the numerator is odd number in this case. To be safe, use "//" which will return the floor value.

File: src/tests/ftest/io/parallel_io.py [hunk 6]
```diff
@@ -192,3 +241,139 @@ def test_parallelio(self):
...
+            pool_job.start()
+        # wait for container create to finish
+        for pool_job in pool_threads:
+            pool_job.join()
+
+        # start dfuse using --svc option only.
+        self.start_dfuse()
+
+        # record free space using statvfs before any data is written.
+        statvfs_info_before = self.statvfs_pool(self.dfuse.mount_dir.value)
```
[the remark this answers is not in the document]
Comment (saurabhtandan) [reply src/tests/ftest/io/parallel_io.py]: Done

File: src/tests/ftest/io/parallel_io.py [hunk 7]
```diff
@@ -192,3 +241,139 @@ def test_parallelio(self):
...
+                cont_num = (pool_count * self.cont_count) + counter
+                dfuse_cont_dir = str(dfuse_pool_dir + "/" +
+                                     self.container[cont_num].uuid)
+                cmd = u"###ls -a {}".format(dfuse_cont_dir)
+                self.execute_cmd(cmd)
+
+                # run ior on all containers
+                test_file = dfuse_cont_dir + "/testfile"
+                self.ior_cmd.test_file.update(test_file)
+                self.ior_cmd.set_daos_params(
```
[the remark this answers is not in the document]
Comment (saurabhtandan) [reply src/tests/ftest/io/parallel_io.py:345]: Addressed in Skype conversation.

File: src/tests/ftest/io/parallel_io.py (hunk 5 above)
Comment (saurabhtandan) [reply src/tests/ftest/io/parallel_io.py]: Done

File: src/tests/ftest/io/parallel_io.py (hunk 4 above)
Comment (saurabhtandan) [reply src/tests/ftest/io/parallel_io.py:389]: Done

File: src/tests/ftest/io/parallel_io.py [hunk 8]
```diff
@@ -102,6 +123,42 @@ def start_dfuse(self):
...
+        except subprocess.CalledProcessError as err:
+            self.fail("Get free block size method failed with: {}".format(err))
+
+        return int(result)
+
+    def statvfs_pool(self, path):
+        """Method to obtain free space using statvfs
+
+          Args:
+            pool_obj (list): List of pool objects.
```
[the remark this answers is not in the document]
Comment (saurabhtandan) [reply src/tests/ftest/io/parallel_io.py]: Done

File: src/tests/ftest/io/parallel_io.py [hunk 9]
```diff
@@ -192,3 +241,139 @@ def test_parallelio(self):
...
+        self.cont_count = self.params.get("cont_count", '/run/container/*')
+        processes = self.params.get("np", '/run/ior/client_processes/*')
+
+        # Create pools in parallel.
+        for _ in range(self.pool_count):
+            pool_thread = threading.Thread(target=self.create_pool())
+            pool_threads.append(pool_thread)
+        # start container create job
+        for pool_job in pool_threads:
+            pool_job.start()
```
[the remark this answers is not in the document]
Comment (saurabhtandan) [reply src/tests/ftest/io/parallel_io.py]: Done

**Comment by saurabhtandan** [comment]
Failed OSAReintegration test failed due to existing issue unrelated to the changes in this PR

File: src/tests/ftest/io/parallel_io.py [hunk 10]
```diff
@@ -192,3 +269,129 @@ def test_parallelio(self):
         # check dfuse is still running after attempting to access deleted
         # container.
             self.dfuse.check_running()
+
+    def test_multipool_parallelio(self):
+        """Jira ID: DAOS-3775.
```
Comment (sylviachanoiyee) [inline src/tests/ftest/io/parallel_io.py:274]: should this be DAOS-3776?

**Comment by sylviachanoiyee** [comment]
the new parallel io aggregation test is set to run on weekly.
do you have a run with the test running and passing using test-tag?

**Comment by saurabhtandan** [comment]
> the new parallel io aggregation test is set to run on weekly.
> do you have a run with the test running and passing using test-tag?

yes Please check the latest test results. It should have a successful run.

**Comment by sylviachanoiyee** [comment]
Thanks Saurabh. Looks good. However, I will open a ticket to get this test and io_aggregation test inside of 'io' to move to 'aggregation' folder.

File: src/tests/ftest/io/parallel_io.py (hunk 10 above)
Comment (saurabhtandan) [reply src/tests/ftest/io/parallel_io.py:274]: As per conversation on Skype, will address it in upcoming ticket.

**Pull request merged**